File: packages/go/whatsapp-go/metering.go
package whatsapp

import (
	"fmt"
	"sync"
	"time"
)

func BillableUnits(used, freeAllowance int64) int64 {
	free := freeAllowance
	if free < 0 {
		free = 0
	}
	out := used - free
	if out < 0 {
		return 0
	}
	return out
}

func BillableDeltaForIncrement(usedBefore, delta, freeAllowance int64) (billableDelta, freeDelta int64) {
	if delta <= 0 {
		before := BillableUnits(usedBefore, freeAllowance)
		after := BillableUnits(usedBefore+delta, freeAllowance)
		return after - before, 0
	}
	before := BillableUnits(usedBefore, freeAllowance)
	after := BillableUnits(usedBefore+delta, freeAllowance)
	billableDelta = after - before
	return billableDelta, delta - billableDelta
}

func EvaluateUsage(used int64, quota CategoryQuota, mode QuotaMode, period string, category MessageCategory, delta int64) MeteringDecision {
	remaining := quota.MonthlyQuota - used
	if remaining < 0 {
		remaining = 0
	}
	overageDelta := used - quota.MonthlyQuota
	if overageDelta < 0 {
		overageDelta = 0
	}
	usedBefore := used - delta
	billableDelta, freeDelta := BillableDeltaForIncrement(usedBefore, delta, quota.FreeAllowance)
	overage := used > quota.MonthlyQuota
	allowed := mode == QuotaSoft || used <= quota.MonthlyQuota
	return MeteringDecision{
		Category:      category,
		Period:        period,
		Used:          used,
		Quota:         quota.MonthlyQuota,
		Remaining:     remaining,
		BillableDelta: billableDelta,
		FreeDelta:     freeDelta,
		Overage:       overage,
		OverageDelta:  overageDelta,
		Allowed:       allowed,
		QuotaMode:     mode,
	}
}

type InMemoryUsageStore struct {
	mu   sync.Mutex
	data map[string]UsageSnapshot
}

func NewInMemoryUsageStore() *InMemoryUsageStore {
	return &InMemoryUsageStore{data: map[string]UsageSnapshot{}}
}

func (s *InMemoryUsageStore) Get(tenantID, period string) (UsageSnapshot, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if found, ok := s.data[tenantID+":"+period]; ok {
		return cloneSnap(found), nil
	}
	return UsageSnapshot{TenantID: tenantID, Period: period, Counts: EmptyUsageCounts()}, nil
}

func (s *InMemoryUsageStore) Increment(tenantID, period string, category MessageCategory, delta int64) (UsageSnapshot, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	key := tenantID + ":" + period
	current, ok := s.data[key]
	if !ok {
		current = UsageSnapshot{TenantID: tenantID, Period: period, Counts: EmptyUsageCounts()}
	}
	current.Counts.Add(category, delta)
	s.data[key] = current
	return cloneSnap(current), nil
}

func cloneSnap(s UsageSnapshot) UsageSnapshot {
	return UsageSnapshot{TenantID: s.TenantID, Period: s.Period, Counts: s.Counts}
}

type MeteringService struct {
	plan     PlanQuotas
	store    UsageStore
	clock    func() time.Time
	timeZone string
}

func NewMeteringService(plan PlanQuotas, store UsageStore, clock func() time.Time, timeZone string) *MeteringService {
	if clock == nil {
		clock = time.Now
	}
	if timeZone == "" {
		timeZone = "America/Sao_Paulo"
	}
	return &MeteringService{plan: plan, store: store, clock: clock, timeZone: timeZone}
}

func (m *MeteringService) Period() string {
	return BillingPeriod(m.clock(), m.timeZone)
}

func (m *MeteringService) Snapshot(tenantID, period string) (UsageSnapshot, error) {
	if period == "" {
		period = m.Period()
	}
	return m.store.Get(tenantID, period)
}

func (m *MeteringService) Consume(tenantID string, category MessageCategory, delta int64) (MeteringDecision, error) {
	if delta <= 0 {
		return MeteringDecision{}, fmt.Errorf("delta de consumo deve ser positivo")
	}
	period := m.Period()
	quota := QuotaFor(m.plan, category)
	snapshot, err := m.store.Increment(tenantID, period, category, delta)
	if err != nil {
		return MeteringDecision{}, err
	}
	used := snapshot.Counts.Get(category)
	decision := EvaluateUsage(used, quota, m.plan.QuotaMode, period, category, delta)
	if !decision.Allowed {
		_, _ = m.store.Increment(tenantID, period, category, -delta)
		return MeteringDecision{}, &QuotaExceededError{
			Message:  fmt.Sprintf("Cota %s excedida no periodo %s (%d/%d, modo hard)", category, period, used, quota.MonthlyQuota),
			Category: category,
			Used:     used,
			Quota:    quota.MonthlyQuota,
			Period:   period,
		}
	}
	return decision, nil
}

func (m *MeteringService) Rollback(tenantID string, category MessageCategory, delta int64) (UsageSnapshot, error) {
	if delta < 0 {
		delta = -delta
	}
	return m.store.Increment(tenantID, m.Period(), category, -delta)
}

File: packages/nodejs/whatsapp-sdk/src/metering.ts
import { QuotaExceededError } from "./errors.js";
import { billingPeriod, emptyUsageCounts, quotaFor } from "./plan.js";
import type {
  MessageCategory,
  MeteringDecision,
  PlanQuotas,
  UsageSnapshot,
  UsageStore,
} from "./types.js";

export function billableUnits(used: number, freeAllowance: number): number {
  return Math.max(0, used - Math.max(0, freeAllowance));
}

/**
 * Quantas unidades do `delta` recém-aplicado são faturáveis, considerando a franquia.
 * Ex.: service usedBefore=950, delta=100, free=1000 → billableDelta=50.
 */
export function billableDeltaForIncrement(
  usedBefore: number,
  delta: number,
  freeAllowance: number,
): { billableDelta: number; freeDelta: number } {
  if (delta <= 0) {
    const billableBefore = billableUnits(usedBefore, freeAllowance);
    const billableAfter = billableUnits(usedBefore + delta, freeAllowance);
    return { billableDelta: billableAfter - billableBefore, freeDelta: 0 };
  }
  const usedAfter = usedBefore + delta;
  const billableBefore = billableUnits(usedBefore, freeAllowance);
  const billableAfter = billableUnits(usedAfter, freeAllowance);
  const billableDelta = billableAfter - billableBefore;
  return { billableDelta, freeDelta: delta - billableDelta };
}

export function evaluateUsage(
  used: number,
  quota: CategoryLike,
  quotaMode: PlanQuotas["quotaMode"],
  period: string,
  category: MessageCategory,
  delta: number,
): MeteringDecision {
  const remaining = Math.max(0, quota.monthlyQuota - used);
  const overageDelta = Math.max(0, used - quota.monthlyQuota);
  const usedBefore = used - delta;
  const { billableDelta, freeDelta } = billableDeltaForIncrement(
    usedBefore,
    delta,
    quota.freeAllowance,
  );
  const overage = used > quota.monthlyQuota;
  const allowed = quotaMode === "soft" || used <= quota.monthlyQuota;
  return {
    category,
    period,
    used,
    quota: quota.monthlyQuota,
    remaining,
    billableDelta,
    freeDelta,
    overage,
    overageDelta,
    allowed,
    quotaMode,
  };
}

interface CategoryLike {
  monthlyQuota: number;
  freeAllowance: number;
}

export class InMemoryUsageStore implements UsageStore {
  private readonly data = new Map<string, UsageSnapshot>();

  constructor(private readonly initial: UsageSnapshot[] = []) {
    for (const snap of initial) {
      this.data.set(keyOf(snap.tenantId, snap.period), cloneSnapshot(snap));
    }
  }

  async get(tenantId: string, period: string): Promise<UsageSnapshot> {
    const found = this.data.get(keyOf(tenantId, period));
    if (found) return cloneSnapshot(found);
    return { tenantId, period, counts: emptyUsageCounts() };
  }

  async increment(
    tenantId: string,
    period: string,
    category: MessageCategory,
    delta: number,
  ): Promise<UsageSnapshot> {
    const current = await this.get(tenantId, period);
    current.counts[category] = Math.max(0, current.counts[category] + delta);
    this.data.set(keyOf(tenantId, period), current);
    return cloneSnapshot(current);
  }
}

function keyOf(tenantId: string, period: string): string {
  return `${tenantId}:${period}`;
}

function cloneSnapshot(snap: UsageSnapshot): UsageSnapshot {
  return {
    tenantId: snap.tenantId,
    period: snap.period,
    counts: { ...snap.counts },
  };
}

export class MeteringService {
  constructor(
    private readonly plan: PlanQuotas,
    private readonly store: UsageStore,
    private readonly clock: () => Date = () => new Date(),
    private readonly timeZone = "America/Sao_Paulo",
  ) {}

  period(at?: Date): string {
    return billingPeriod(at ?? this.clock(), this.timeZone);
  }

  async snapshot(tenantId: string, period?: string): Promise<UsageSnapshot> {
    return this.store.get(tenantId, period ?? this.period());
  }

  async consume(
    tenantId: string,
    category: MessageCategory,
    delta = 1,
  ): Promise<MeteringDecision> {
    if (delta <= 0) {
      throw new Error("delta de consumo deve ser positivo");
    }
    const period = this.period();
    const quota = quotaFor(this.plan, category);
    const snapshot = await this.store.increment(tenantId, period, category, delta);
    const used = snapshot.counts[category];
    const decision = evaluateUsage(used, quota, this.plan.quotaMode, period, category, delta);
    if (!decision.allowed) {
      await this.store.increment(tenantId, period, category, -delta);
      throw new QuotaExceededError(
        `Cota ${category} excedida no periodo ${period} (${used}/${quota.monthlyQuota}, modo hard)`,
        { category, used, quota: quota.monthlyQuota, period },
      );
    }
    return decision;
  }

  async rollback(tenantId: string, category: MessageCategory, delta = 1): Promise<UsageSnapshot> {
    const period = this.period();
    return this.store.increment(tenantId, period, category, -Math.abs(delta));
  }
}
